### PR TITLE
feat(assets): add palette effect system with cycling, fade, flash, and swap

### PR DESCRIPTION
**Pull request by vancura**
Introduce PaletteEffectManager that runs after demo.render() and before GPU upload, with an activeCount guard to skip work when idle.

Effects:
- CycleEffect: constant-speed palette rotation (zero-alloc)
- FadeEffect/FadeRangeEffect: timed interpolation with easing
- FlashEffect: temporary color override with auto-restore
- paletteSwap: instant two-entry exchange

Add getRef() and markDirty() to Palette for zero-copy mutation. Expose all effects on the BT public facade (6 new methods). Includes 31 unit tests and Tier 1 CPU benchmarks.

## Overview

This PR adds a palette effects system to the BlitTech engine, enabling runtime animation and manipulation of color palettes. The system runs in the render loop after demo.render() and before GPU upload, and uses an activeCount guard to skip processing when idle. In-flight effects are cleared when the active palette is replaced to avoid stale snapshots.

## Core Components

**Palette API Enhancements** (`src/assets/Palette.ts`):
- Added `getRef(index)` for zero-copy direct access to internal `Color32` entries.
- Added `markDirty()` to explicitly flag the palette for renderer upload on the next frame.

**Palette Effects System** (`src/assets/PaletteEffect.ts`):
- `PaletteEffect` interface: `update(palette, deltaMs): boolean`.
- `PaletteEffectManager`:
  - Injectable time provider (defaults to `performance.now()`).
  - Per-frame delta time computation; first-frame delta is skipped (treated as 0) to avoid large accumulated gaps.
  - In-place compaction/removal of completed effects (no reallocation).
  - `activeCount` getter; adding the first effect after idle resets the manager clock to prevent a large delta spike.
  - `add(effect)`, `update(palette)`, `clear()`.
- Effects implemented:
  - `CycleEffect`: Continuous index-range rotation at configurable steps/sec, fractional accumulator, preallocated temp buffer for zero/allocation rotation.
  - `FadeEffect`: Full-palette timed interpolation with optional easing; snapshots source and converges exactly to target.
  - `FadeRangeEffect`: Range-limited fade affecting indices `[start, end]` only.
  - `FlashEffect`: Temporary color override of non-zero entries with automatic restore from snapshot after duration.
- `paletteSwap(palette, indexA, indexB)`: Instant two-entry in-place exchange; marks palette dirty.

All effects preserve transparent index 0 where applicable.

## Public API Additions

**`BT` facade** (`src/BlitTech.ts`):
- Exposed methods delegating to `BTAPI.instance`: `paletteCycle(start, end, speed)`, `paletteFade(target, durationMs, easing?)`, `paletteFadeRange(start, end, target, durationMs, easing?)`, `paletteFlash(color, durationMs)`, `paletteSwap(indexA, indexB)`, `paletteClearEffects()`.

**`BTAPI` class** (`src/core/BTAPI.ts`):
- New private field: `paletteEffects: PaletteEffectManager`.
- Integrated `paletteEffects.update(this.palette)` into the render path (after demo.render(), before renderer endFrame()), executed only when `paletteEffects.activeCount > 0` and an active palette exists.
- `setPalette(...)` now clears in-flight palette effects (`paletteEffects.clear()`) when replacing the active palette to prevent effects from applying to stale palettes.
- Added public methods implementing the palette-effect APIs and a private `assertFiniteDuration` helper for consistent duration validation.

## Validation & Behavior Notes

- Input validation added to palette-effect APIs:
  - `paletteCycle`: requires finite `speed` and integer `start < end`.
  - `paletteFade`, `paletteFadeRange`, `paletteFlash`: require finite, non-negative `durationMs`.
  - `paletteSwap`: requires an active palette; no-op if indices are identical.
- `paletteFlash` now requires an active palette before queuing (fix): ensures flashes are only queued when a palette exists to avoid invalid/no-op queueing.
- When the manager transitions from idle to active (first effect added), its internal clock is reset to avoid a large first-frame delta.
- Palette dirtiness is marked after effect updates to ensure GPU sync.
- Effects auto-remove by returning `false` from `update()` when complete.

## Testing & Benchmarks

- Unit tests: 31 tests covering manager lifecycle, effect behaviors and edge cases (zero duration, identical indices, out-of-range), easing modes, preservation of index 0, and concurrent effects (`src/assets/PaletteEffect.test.ts`).
- Benchmarks: Tier 1 CPU benchmarks for `CycleEffect.update()`, `FadeEffect.update()`, `paletteSwap()`, and `PaletteEffectManager.update()` (active vs idle scenarios) (`src/assets/PaletteEffect.bench.ts`).

## Implementation Notes

- Designed for zero/allocation hotspots: cycle rotations use preallocated buffers and manager uses in-place compaction.
- Effects snapshot source/target data as needed to ensure deterministic interpolation and correct restoration for flash.
- `getRef()` + `markDirty()` enable zero-copy external mutations while requiring explicit dirty signaling for renderer uploads.

### CODE REVIEW

File: src/assets/Palette.ts [hunk 1]
```diff
@@ -336,6 +336,28 @@ export class Palette {
...
+     * {@link get} and {@link set} to keep the dirty flag accurate automatically.
+     *
+     * @param index - Palette index to access.
+     * @returns Direct reference to the internal color entry.
+     * @throws Error if the index is invalid.
+     */
+    public getRef(index: number): Color32 {
+        this.assertIndexInRange(index);
+
+        return this.colorAt(index);
```
[the remark this answers is not in the document]
Comment (vancura) [reply src/assets/Palette.ts:358]: @coderabbitai: Not fixing. All effect implementations already skip index 0 in their loops. Adding a branch to getRef would add per-call overhead in hot loops for an invariant that's already enforced at the caller level.

File: src/assets/PaletteEffect.ts [hunk 2]
```diff
@@ -0,0 +1,464 @@
...
+                if (saved) {
+                    palette.getRef(i).copyFrom(saved);
+                }
+            }
+
+            return false;
+        }
+
+        // Keep flashing.
+        return true;
```
[the remark this answers is not in the document]
Comment (vancura) [reply src/assets/PaletteEffect.ts:427]: @coderabbitai: Not fixing. Flash applies once and holds; the palette stays flashed because nothing overwrites it between frames. Adding per-frame re-writes and conditional restores would make behavior unpredictable when combined with other effects.

File: src/core/BTAPI.ts [hunk 3]
```diff
@@ -196,6 +209,14 @@ export class BTAPI {
                 if (this.renderer) {
                     this.renderer.beginFrame();
                     this.demo?.render();
+
+                    // Palette effects run after demo render (so user's explicit palette
+                    // changes in render() are respected) but before endFrame (so effects
+                    // are visible this frame via the dirty-flag GPU upload).
+                    if (this.palette && this.paletteEffects.activeCount > 0) {
+                        this.paletteEffects.update(this.palette);
+                    }
```
Comment (vancura) [inline src/core/BTAPI.ts:218]: @coderabbitai: Not fixing. Palette.getRef() already calls assertIndexInRange() which throws on out-of-bounds access. Adding redundant checks at the BTAPI layer is unnecessary. Index 0 is valid for cycling and swapping.

[the remark this answers is not in the document]
Comment (vancura) [reply src/core/BTAPI.ts:218]: @coderabbitai: Not fixing. The suggestion to keep stale effects running across palette changes is wrong. `FadeEffect` and `FlashEffect` hold snapshot `Color32` arrays cloned from the old palette. If the palette changes, those snapshots become meaningless - the fade would lerp toward colors that have nothing to do with the new palette, and the flash would restore the old palette's colors on top of the new one. Clearing is the correct behavior. The previous review round identified this exact bug and we fixed it.

File: src/core/BTAPI.ts [hunk 4]
```diff
@@ -308,6 +329,10 @@ export class BTAPI {
             console.warn('[BT] Active palette structure changed. Call BT.spritesRefresh() to update loaded sprites.');
         }
 
+        // In-flight effects hold snapshots of the old palette. Drop them so they
+        // don't apply stale colors to the new palette.
+        this.paletteEffects.clear();
```
[the remark this answers is not in the document]
Comment (vancura) [reply src/core/BTAPI.ts:334]: @coderabbitai: Already fixed. The `add()` method resets `lastTime = 0` when transitioning from empty to non-empty (line 70-72). Then `update()` checks `lastTime === 0` and returns `delta=0` for that first call (line 87). The guard at line 216 is fine - when activeCount is 0, there's nothing to update, and the clock resets on the next `add()`. Both mechanisms work together correctly.
